Fix mold_indicator errors at startup

ehendrix23 opened the pull request:
## Description:

See to resolve issue as reported in 16733 on startup. 

Added logic to ensure that if the state is unknown during startup that the error about being unable to parse the value is not logged.
Further,  also ensured that if an attribute is set to None it does not try to convert the None value to Fahrenheit as that will cause an error.

**Related issue (if applicable):** fixes #16733

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

ehendrix23 committed:
Initial changes to resolve issue 16733

Added logic to ensure that if the state is unknown during startup that the error about being unable to parse the value is not logged.
Further,  also ensured that if an attribute is set to None it does not try to convert the None value to Fahrenheit as that will cause an error.

ehendrix23 committed:
Cleaned up and added few comments

Cleaned up some lines based on flake8 and pylint.
Added some comment lines on the items added.

MartinHjelmare reviewed (commented):
Move the first state update and state tracking registration to home assistant start event and register that listener from `async_added_to_hass`.

ehendrix23 committed:
Changed to async and using async_added_to_hass

Changed sensor to use async.
Registering state tracking for sensors and initial setup is now done upon the home assistant start event.

ehendrix23 commented:
@MartinHjelmare I believe I just completed what you mentioned. And as part of it also changed the sensor to async instead. This is first time doing this, appreciate you checking (which I know you will <g>) and providing feedback. :-)

Thank you!!!

ehendrix23 committed:
Updated test and small fix

Updated test to handle unavailable state of sensor and return of None for attributes when data is unavailable.
Ensured that atributes are set to None when state is unavailable due to incorrect data.

MartinHjelmare commented on homeassistant/components/sensor/mold_indicator.py:
```diff
@@ -41,66 +44,125 @@
-        indoor_humidity_sensor, calib_factor)], True)
+        indoor_humidity_sensor, calib_factor)], False)
+
+    return True
```
Nothing is checking this return value.

MartinHjelmare commented on homeassistant/components/sensor/mold_indicator.py:
```diff
@@ -41,66 +44,125 @@
 
-    def __init__(self, hass, name, indoor_temp_sensor, outdoor_temp_sensor,
-                 indoor_humidity_sensor, calib_factor):
+    def __init__(self, hass, name, indoor_temp_sensor,
```
Pass in `is_metric` instead of `hass`. `hass` will be set on the entity when it has been added to home assistant.

MartinHjelmare replied:
Most often it's a bad pattern yes. But `hass` will not be set until the entity has been added. So we can't use `self.hass` before that. It's ok to use `self.hass` in `update` and `async_added_to_hass` but not in `__init__`.

MartinHjelmare commented on homeassistant/components/sensor/mold_indicator.py:
```diff
@@ -41,66 +44,125 @@
+    def __init__(self, hass, name, indoor_temp_sensor,
+                 outdoor_temp_sensor, indoor_humidity_sensor, calib_factor):
         """Initialize the sensor."""
+        self.hass = hass
```
Remove this.

ehendrix23 replied:
Done

MartinHjelmare commented on homeassistant/components/sensor/mold_indicator.py:
```diff
@@ -41,66 +44,125 @@
+            outdoor_temp = self.hass.states.get(self._outdoor_temp_sensor)
+            indoor_hum = self.hass.states.get(self._indoor_humidity_sensor)
+
+            schedule_update = True if self._update_sensor(
```
```py
schedule_update = self._update_sensor(...)
```

ehendrix23 replied:
Was trying to be too smart for my own good here. :-) Changed.

ehendrix23 committed:
Fixed some flake8 issues in test

Fixed some flake8 issues in test_moldindicator.py.

ehendrix23 committed:
Updates based on review

Updates based on review from MartinHjelmare

MartinHjelmare commented on tests/components/sensor/test_moldindicator.py:
```diff
@@ -98,6 +147,66 @@ def test_calculation(self):
+        }))
+        self.hass.start()
+
+        self.hass.states.set('test.indoortemp', 'UNKNOWN',
```
The string for `UNKNOWN_STATE` is `'unknown'`.

ehendrix23 committed:
Added sensor entity_id to logger errors

Added sensor entity_id to logger error messages
Update test to use constant STATE_UNKNOWN instead of fixed string.

MartinHjelmare reviewed (approved):
Looks good!
